File: Dynamic Programming/UniquePath3.js
/**
 * Problem: https://leetcode.com/problems/unique-paths-iii/submissions/
 * Author: Dung Nguyen Tien (Chris)
 * @param {number[][]} grid
 * @return {number}
 */
var uniquePathsIII = function(grid) {
    var start = [];
    var end = [];
    var walks = 2;
    var steps = [];
    res = [];
    
    for (var i = 0; i < grid.length; i++) {
        for (var j = 0; j < grid[0].length; j++) {
            if (grid[i][j] == 1) start = [i, j];
            else if (grid[i][j] == 2) end = [i, j];
            else if (grid[i][j] == 0) walks++;
        }
    }
    
    // find all path and push the value to res
    steps.push(start);
    dfs(grid, start, end, walks, steps);

    return res.length;
};
var res;

function dfs(grid, position, end, walks, steps) {
    // if reach the end point, return
    if (position[0] == end[0] && position[1] == end[1] && steps.length == walks) {
        res.push(steps);
        return
    }
    
    // check if the step go all the empty square
    if (steps.length < walks) {
        var dir = [[0, 1], [1,0], [0, -1], [-1, 0]];
        
        for (var i = 0; i < dir.length; i++) {
            var nextX = position[0] + dir[i][0];
            var nextY = position[1] + dir[i][1];
            var nextMove = [nextX, nextY];
            
            // check if can make the next move
            if (nextX >= 0 && nextX < grid.length 
                && nextY >= 0 && nextY < grid[0].length 
                && grid[nextX][nextY] != -1
                && !steps.some(s => s[0] == nextX && s[1] == nextY)) {
                
                // add the next move
                steps.push(nextMove);
                dfs(grid, nextMove, end, walks, steps);
                // after find or not find the path, remove the nextMove the add another one to move on
                steps.pop()
            }
        }
    }
    return true;
}

File: Dynamic Programming/UniquePath3.cs
/*
 * Problem: https://leetcode.com/problems/unique-paths-iii/
 * Author: Dung Nguyen Tien (Chris)
*/

public class Solution {
    public int UniquePathsIII(int[][] grid) {
        var rows = grid.Length;
        var cols = grid[0].Length;
        var walk = 2;
        var start = new int[] { -1, -1 };
        var end = new int[] { -1, -1 };
        
        for (var i = 0; i < rows; i++) {
            for (var j = 0; j < cols; j++) {
                if (grid[i][j] == 1) start = new int[] { i, j };
                else if (grid[i][j] == 2) end = new int[] { i, j };
                else if (grid[i][j] == 0) walk++;
            }
        }
        
        var result = new List<List<int[]>>();
        var steps = new List<int[]>();
        steps.Add(start);
        
        DFS(grid, start, end, result, steps, walk);
        
        return result.Count;
    }
    
    public void DFS(int[][] grid, int[] current, int[] end, List<List<int[]>> result, List<int[]> steps, int walk) 
    {
        if (ComparePosition(current, end) && steps.Count == walk) {
            result.Add(new List<int[]>(steps));
            return;
        }
        
        var directions = new int[][] {
            new int[] { 1, 0 },
            new int[] { -1, 0 },
            new int[] { 0, 1 },
            new int[] { 0, -1 }
        };
        
        foreach(var d in directions) {
            var nextMove = new int[2];
            nextMove[0] = current[0] + d[0]; 
            nextMove[1] = current[1] + d[1];
            
            if (nextMove[0] < 0 || nextMove[0] >= grid.Length
               || nextMove[1] < 0 || nextMove[1] >= grid[0].Length
               || grid[nextMove[0]][nextMove[1]] == -1
               || steps.Any(step => ComparePosition(step, nextMove))) {
                continue;
            }
            
            steps.Add(nextMove);
            DFS(grid, nextMove, end, result, steps, walk);
            steps.RemoveAt(steps.Count - 1);
        }
    }
    
    public bool ComparePosition(int[] p1, int[] p2) {
        return p1[0] == p2[0] && p1[1] == p2[1];
    }
}